qiskit/optimization/converters/optimization_problem_to_negative_value_oracle.py

```diff
@@ -0,0 +1,182 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""OptimizationProblemToNegativeValueOracle module"""
+
+import logging
+import numpy as np
+from typing import Optional, Tuple, Dict, Union
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, Aer
+from qiskit.aqua.components.oracles import CustomCircuitOracle
+from qiskit.aqua.components.initial_states import Custom
+from qiskit.aqua.components.iqfts import Standard as IQFT
+from qiskit.providers import BaseBackend
+from qiskit.optimization.problems import OptimizationProblem
+
+
+class OptimizationProblemToNegativeValueOracle:
+
+    """Converts an optimization problem (QUBO) to a negative value oracle.
+
+    In addition, a state preparation operator is generated from the coefficients and constant of a
+    QUBO, which can be used to encode the function into a quantum state. In conjuction, this oracle
+    and operator can be used to flag the negative values of a QUBO encoded in a quantum state.
+    """
+
+    def __init__(self, num_output_qubits: int, backend: Optional[BaseBackend] = None) -> None:
+        """
+        Args:
+            num_output_qubits: The number of qubits required to represent the output.
+            backend: Instance of selected backend.
+        """
+        self._num_key = 0
+        self._num_value = num_output_qubits
+        if backend is None:
+            self._backend = Aer.get_backend('statevector_simulator')
+        else:
+            self._backend = backend
+        self._logger = logging.getLogger(__name__)
+
+    def encode(self, problem: OptimizationProblem) -> \
+            Tuple[Custom, CustomCircuitOracle, Dict[Union[int, Tuple[int, int]], int]]:
+        """ A helper function that converts a QUBO into an oracle that recognizes negative numbers.
+
+        Args:
+            problem: The problem to be solved.
+
+        Returns:
+            A state preparation operator A, an oracle O that recognizes negative numbers, and
+            a dictionary representation of the function coefficients, where the key -1 represents
+            the constant.
+        """
+
+        # get linear part of objective
+        linear_dict = problem.objective.get_linear()
+        linear_coeff = np.zeros(problem.variables.get_num())
+        for i, v in linear_dict.items():
+            linear_coeff[i] = v
+
+        # get quadratic part of objective
+        quadratic_dict = problem.objective.get_quadratic()
+        quadratic_coeff = {}
+        for i, jv in quadratic_dict.items():
+            for j, v in jv.items():
+                coeff = quadratic_coeff.get((j, i), 0)
+                if i <= j:
+                    # divide by 2 since problem considers xQx/2.
+                    quadratic_coeff[(i, j)] = v / 2 + coeff
+                else:
+                    quadratic_coeff[(j, i)] = v / 2 + coeff
```

Comment:
I think `jv` will raise a pylint error since it doesn't comply with the naming convention (of at least 3 characters), but that's quickly fixed
```suggestion
        for i, j_value_dict in quadratic_dict.items():
            for j, value in j_value_dict.items():
                coeff = quadratic_coeff.get((j, i), 0)
                if i <= j:
                    # divide by 2 since problem considers xQx/2.
                    quadratic_coeff[(i, j)] = value / 2 + coeff
                else:
                    quadratic_coeff[(j, i)] = value / 2 + coeff
```